File: clients/ruby/lib/haveapi/client/authentication/basic.rb
require 'haveapi/client/authentication/base'

module HaveAPI::Client::Authentication
  class Basic < Base
    register :basic

    def resource
      RestClient::Resource.new(
        @communicator.url,
        user: @opts[:user],
        password: @opts[:password],
        verify_ssl: @communicator.verify_ssl
      )
    end

    def user = @opts.[](:user)
    def password = @opts.[](:password)
  end
end

File: clients/elixir/lib/haveapi/client/authentication/basic.ex
defmodule HaveAPI.Client.Authentication.Basic do
  alias HaveAPI.Client

  @behaviour Client.Authentication

  def setup(_conn, opts) do
    {:ok, %{username: opts[:username], password: opts[:password]}}
  end

  def authenticate(req, opts) do
    Client.Request.add_header(
      req,
      "Authorization",
      "basic " <> Base.encode64("#{opts.username}:#{opts.password}")
    )
  end

  def logout(_conn, _opts), do: :ok
end
